packages/framework/src/Component/EntityExtension/EntityExtensionSubscriber.php

```diff
@@ -0,0 +1,306 @@
+<?php
+
+namespace Shopsys\FrameworkBundle\Component\EntityExtension;
+
+use Doctrine\Common\EventSubscriber;
+use Doctrine\ORM\Configuration;
+use Doctrine\ORM\Event\LoadClassMetadataEventArgs;
+use Doctrine\ORM\Events;
+use Doctrine\ORM\Mapping\ClassMetadataInfo;
+use LogicException;
+use ReflectionClass;
+use Webmozart\Assert\Assert;
+
+/**
+ * Inspired by joschi127/doctrine-entity-override-bundle (https://github.com/joschi127/doctrine-entity-override-bundle)
+ */
+class EntityExtensionSubscriber implements EventSubscriber
+{
+    /**
+     * @var array<class-string, class-string>
+     */
+    protected array $entityExtensionMap;
+
+    /**
+     * @var bool
+     */
+    protected bool $isEntityMapLoaded = false;
+
+    /**
+     * @var array<class-string, array<class-string, class-string>>
+     */
+    protected array $parentClassesByClass = [];
+
+    /**
+     * @var array<class-string, class-string>
+     */
+    protected array $allParentClasses = [];
+
+    /**
+     * @var class-string[]|null
+     */
+    protected ?array $allRegisteredEntitiesCache = null;
```

Comment:
TBH, I am not a big fan of nullable arrays. I find an empty array more sexy than `null`. Is there a particular reason why it is nullable?

Reply:
In this case, it distinguishes `not-loaded state` and `loaded empty state`. 

In practice, it's unexpected that there will be no registered entity, so If you find it more understandable, I can change the nullability and load registered entities on an empty array.

Reply:
Nah, you can keep it as is, the explanation is good enough for me, thank you :slightly_smiling_face: